modules/create_timepoint/jsx/createTimepointIndex.js

```diff
@@ -0,0 +1,358 @@
+        }
+        // Populate the select options for subproject.
+        if (data.hasOwnProperty('subproject')) {
+          state.form.options.subproject = data.subproject;
+          state.form.value.subproject = null;
+          state.form.display.subproject = true;
+        }
+        // Populate the select options for psc.
+        if (data.psc) {
+          state.form.options.psc = data.psc;
+          state.form.value.psc = null;
+          state.form.display.psc = true;
+        }
+        // Populate the select options for visit.
+        if (data.visit) {
+          // Store the (complete) visit selection information.
+          state.storage.visit = data.visit;
+          // Handle visit selection.
+          this.handleVisitLabel();
+        }
+        // Populate the select options for languages.
+        if (data.languages) {
+          state.form.options.languages = data.languages;
+          state.form.value.languages = null;
+          state.form.display.languages = true;
+        }
+        this.setState(state);
+      }).catch((error) => {
+        const state = Object.assign({}, this.state);
+        state.errors = [{message: error}];
+        this.setState(state);
+      });
+  }
+
+  /**
+   * Visit Labels refreshes when Subproject changes.
+   */
+  handleVisitLabel() {
+    const state = Object.assign({}, this.state);
+    state.form.options.visit = state.storage.visit[
+      state.form.value.subproject
+      ];
+    if (state.form.value.subproject) {
+      state.form.value.visit = Object.keys(state.storage.visit[
+        state.form.value.subproject
+        ])[0];
+    }
+    state.form.display.visit = true;
+    this.setState(state);
+  }
+
+  /**
+   * Set the form data based on state values of child elements/components
+   *
+   * @param {string} formElement - name of the selected element
+   * @param {string} value - selected value for corresponding form element
+   */
+  setForm(formElement, value) {
+    const state = Object.assign({}, this.state);
+    state.form.value[formElement] = value;
+    this.setState(state);
+    if (formElement === 'subproject') {
+      this.handleVisitLabel();
+    }
+  }
+
+  /**
+   * Handle form submission
+   * @param {object} e - Form submission event
+   */
+  handleSubmit(e) {
+    e.preventDefault();
+    const state = Object.assign({}, this.state);
+    const url = `${this.props.baseURL}/create_timepoint/Timepoint`;
+    const send = JSON.stringify({
+      candID: state.url.params.candID,
+      identifier: state.url.params.identifier,
+      subproject: state.form.value.subproject,
+      psc: state.form.value.psc,
+      visit: state.form.value.visit,
+      languages: state.form.value.languages,
+    });
+    fetch(
+      url, {
+        method: 'POST',
+        credentials: 'same-origin',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: send,
+      }).then((response) => {
+      return response.json();
```

Comment:
i think it would be nice here to add a success swal if (response.ok) or (response.status === 200). if not a swal, at least something to replace the current confirmation 'New time point successfully registered':

<img width="685" alt="Screen Shot 2020-07-07 at 6 31 56 PM" src="https://user-images.githubusercontent.com/17415878/86852072-1511f000-c082-11ea-85da-3312a0f7f1ad.png">

i also think it would be better to check if response.ok first, then else, call response.json() and check if there errors. this way, you are checking against the response actually being ok, rather than whether the response returns an error in the json.

```
    .then((response) => {
      if (response.ok || response.status === 200) {
        swal.fire('Success!', 'New time point successfully registered.', 'success').then((result) => {
          if (result.value) {
              // whatever you need to do here i.e. refresh / replace
          }
        });
      } else {
        response.json().then((data) => {
          if (data.errors) {
             // do what you want to do here
          }
        });
      }
    })
    .catch((error) => {
      console.error(error);
    });
```

Reply:
@zaliqarosli great idea. I'll add it.

Reply:
@zaliqarosli I'm assuming but I think if the `return response.json()` were to throw an error the following code would run:
```
.catch((error) => {
        console.error('Error! ' + error);
      });
```
I can change it in any case.

Reply:
hey @maltheism , sorry i'm only just seeing this comment now. you would still access the custom error messages using `data.errors` in the else block i.e.

```
        response.json().then((data) => {
          if (data.errors) {
             // access custom errors messages here with data.errors
             // do what you want to do here
          }
        });
```

i'm not sure what you mean by 

> That would be the same as response.ok or response.status being an error

response.ok and response.status === 200 is by definition a response without any errors. the catch error block you pasted above would be called i think only if there was a 500 or a bug with the component. as long as you don't have any JS or PHP errors, the `.then((response) => {` block would be entered and THEN, you have to see if the response is OK or maybe a 409 or whatever.

response.json() would never throw an response error because what response.json() is is just some string containing the custom error message that you then use to display. but receiving the response.json() error string isn't a response error itself, it actually is a 200 OK "I have received this string"

i found this little bit from https://reactjs.org/docs/faq-ajax.html helpful. let me know if you want to discuss further on slack!

<img width="582" alt="Screen Shot 2020-07-20 at 11 34 25 AM" src="https://user-images.githubusercontent.com/17415878/87956824-7ae87980-ca7d-11ea-8678-3efc36a9710d.png">